README.rst

```diff
@@ -1,22 +1,230 @@
+that could be done.
+
+Let's complicate the example a bit further and assume that the bug was
+introduced in `bash-4.4.16-1.fc27` and was fixed in
+`bash-4.4.22-1.fc27`.  Now we would need to compare the version in a
+range of NVRs making the script quite a bit more complicated.
+Especially if we wanted to be able to apply it to other packages in the
+future and the parsing of the NVR and version comparison would need to
+be generalized.
+
+So, how would we code this using `insights-core`?  The following script
+will identify whether the current system has a version of `bash`
+installed within the range.
+
+::
+
+    from insights import rule, make_response, run
+    from insights.parsers.installed_rpms import InstalledRpms, InstalledRpm
+
+    lower = InstalledRpm.from_package("bash-4.4.16-1.fc27")
+    upper = InstalledRpm.from_package("bash-4.4.22-1.fc27")
+
+    @rule(InstalledRpms)
+    def report(rpms):
+        rpm = rpms.get_max("bash")
+        if rpm and rpm >= lower and rpm < upper:
+            return make_response("BASH_AFFECTED", version=rpm.nvr)
+        elif rpm:
+            return make_response("BASH_UNAFFECTED", version=rpm.nvr)
+        else:
+            return make_response("NO_BASH_FOUND")
+
+    if __name__ == "__main__":
+        run(report, print_summary=True)
+
+To keep this preview simple, we won't unpack everything here, delegating
+to the other tutorials and examples for further details.  However, let's
+highlight some points to help understand what's going on.
+
+After importing the components we need, we define the bounds of our
+version range by creating a `lower` and `upper` `InstalledRpm` version.
+Then we define our `rule` which depends upon the `InstalledRpms`
+"parser" class.  This class will be passed in as the first argument to
+the `report` function.
+
+The `report` function gets the maximum version of `bash` found in
+`InstalledRpms` and then compares it to the `upper` and `lower` values.
+If a version of `bash` is found between those versions, it returns
+`BASH_AFFECTED` and the version it found.  Otherwise, if it finds a
+version of `bash` that is outside the range, it returns
+`BASH_UNAFFECTED` and the version found.  Finally, if it doesn't find
+any version of bash, it returns `NO_BASH_FOUND`.
+
+This is all executed by the `run` command.  The `run` command will
+"know" about the `report` function and that it depends upon
+`InstalledRpms` based on the `@rule` decorator.  
+
+That's cool, but is it really simpler?
+--------------------------------------
+
+Certainly, for a simple check, just grepping a file is simpler.
+However, there are gains as the complexity and automation requirements
+come into play.  In this particular case, the complexity is hidden in two
+places.
+
+First, complexity around parsing the `rpm -qa` command, and presenting
+it in a usable data model is handled by the `InstalledRpms` class.  It
+is just one(link) of many parsers(link) and combiners(link) that one can use.   In
```

Comment:
Missing URL's?

Reply:
Dangit.  Yes, those were reminders that didn't remind me.